pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/decoder/protobufnative/PulsarProtobufNativeRowDecoderFactory.java

```diff
@@ -0,0 +1,167 @@
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.TypeManager;
+import io.prestosql.spi.type.TypeSignature;
+import io.prestosql.spi.type.TypeSignatureParameter;
+import io.prestosql.spi.type.VarbinaryType;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.pulsar.client.impl.schema.generic.GenericProtobufNativeSchema;
+import org.apache.pulsar.common.naming.TopicName;
+import org.apache.pulsar.common.schema.SchemaInfo;
+import org.apache.pulsar.sql.presto.PulsarColumnHandle;
+import org.apache.pulsar.sql.presto.PulsarColumnMetadata;
+import org.apache.pulsar.sql.presto.PulsarRowDecoder;
+import org.apache.pulsar.sql.presto.PulsarRowDecoderFactory;
+
+/**
+ * PulsarRowDecoderFactory for {@link org.apache.pulsar.common.schema.SchemaType#PROTOBUF_NATIVE}.
+ */
+public class PulsarProtobufNativeRowDecoderFactory implements PulsarRowDecoderFactory {
+
+    private TypeManager typeManager;
+
+    public PulsarProtobufNativeRowDecoderFactory(TypeManager typeManager) {
+        this.typeManager = typeManager;
+    }
+
+    @Override
+    public PulsarRowDecoder createRowDecoder(TopicName topicName, SchemaInfo schemaInfo,
+                                             Set<DecoderColumnHandle> columns) {
+        return new PulsarProtobufNativeRowDecoder((GenericProtobufNativeSchema) GenericProtobufNativeSchema.of(schemaInfo), columns);
+    }
+
+    @Override
+    public List<ColumnMetadata> extractColumnMetadata(TopicName topicName, SchemaInfo schemaInfo,
+                                                      PulsarColumnHandle.HandleKeyValueType handleKeyValueType) {
+        List<ColumnMetadata> columnMetadata;
+        String schemaJson = new String(schemaInfo.getSchema());
+        if (StringUtils.isBlank(schemaJson)) {
+            throw new PrestoException(NOT_SUPPORTED, "Topic "
+                    + topicName.toString() + " does not have a valid schema");
+        }
+        Descriptors.Descriptor schema;
+        try {
+            schema =
+                    ((GenericProtobufNativeSchema) GenericProtobufNativeSchema.of(schemaInfo)).getProtobufNativeSchema();
+        } catch (Exception ex) {
+            throw new PrestoException(NOT_SUPPORTED, "Topic "
+                    + topicName.toString() + " does not have a valid schema");
+        }
+
+        try {
+            columnMetadata = schema.getFields().stream()
+                    .map(field ->
+                            new PulsarColumnMetadata(PulsarColumnMetadata.getColumnName(handleKeyValueType, field.getName()),
+                                    parseProtobufPrestoType(field), field.getType().toString(), null, false, false,
+                                    handleKeyValueType, new PulsarColumnMetadata.DecoderExtraInfo(field.getName(),
+                                    null, null))
+
+                    ).collect(toList());
+        } catch (StackOverflowError e) {
```

Comment:
Why should we catch this exception ?
IMHO this is a code smell, if we fall into a StackOverflowError then there is a bad problem somewhere else

Reply:
Schema cyclic definitions which may case java.lang.StackOverflowError , also cyclic definitions is bad schema in relational database ( Ref spark-sql : [JavaTypeInference](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala#142)) ) . But seemingly , Protobuf have not support Cyclic Objects (https://github.com/protocolbuffers/protobuf/issues/5504) .
Thank you for your suggest , I will remove the `try` code.